### DISCUSSION TITLE
ParticleEmitter leaving particles in scene after being killed

### DISCUSSION THREAD

**Discussion by kyletcooper** [excaliburjs/Excalibur#3254] [Q&A] [answered]
I'm having an issue using the ParticleEmitter in my endless wave game. As the rounds progress the performance is getting worse and I noticed that in engine.currentScene.entities contains a lot of particles that are never getting removed.

```
onInitialize() {
	this.addChild(
		new ParticleEmitter({
			radius: 5,
			emitterType: EmitterType.Circle,
			minAngle: Math.PI / 2,
			maxAngle: Math.PI / 2,
			isEmitting: true,
			emitRate: 20,
			opacity: 0.75,
			fadeFlag: true,
			particleLife: 250,
			startSize: Bullet.size,
			endSize: Bullet.size,
			beginColor: Palette.Bullet(),
		})
	);

	this.spawnSubShots();
}

onPostKill(scene: Scene): void {
	const emitter = new ParticleEmitter({
		pos: this.pos,
		radius: 2,

		minSize: 1,
		maxSize: 2,
		startSize: 2,
		endSize: 0,

		minVel: 100,
		maxVel: 200,

		minAngle: toRadians(-270),
		maxAngle: toRadians(0),

		isEmitting: true,
		emitRate: 120,

		particleLife: 30,

		beginColor: Palette.Bullet(),
		endColor: Palette.Powerup(),
	});

	scene.add(emitter);

	scene.engine.clock.schedule(() => {
		emitter.kill();
	}, 100);
}
```

Here's what I think my code is doing:

1. **Impact** - My idea is that I create an emitter at the location and then kill it myself after a short delay.
2. **Trail** - Here I attach an emitter to the bullet so as it moves it leaves particles behind. Again my understanding is that this gets killed/removed whenever the parent bullet does.

Is the ParticleEmitter meant to remove it's particles when it's killed or am I misusing this? I had assumed that the particleLife property was used to have the particles automatically remove themselves after they're finished.

**Comment by eonarheim** [comment]
Hi @kyletcooper! Apologies on the delay I'm catching up from last week.

I don't see anything that looks wrong in the code. However there was a memory leak in v0.29.3 particles that would definitely contribute to perf degradation, In theory the latest alpha bits have fixed this by correctly pooling particles. 

> Is the ParticleEmitter meant to remove it's particles when it's killed or am I misusing this? I had assumed that the particleLife property was used to have the particles automatically remove themselves after they're finished.

This is a good question, currently the ParticleEmitter does not remove spawned particles when removed from the scene (perhaps it should, I'll look into making this change). It does look like there is a bug where particles don't clean up properly if their emitter is removed before their life runs out 😱 I'll make an issue and get this fixed ASAP

**Reply by eonarheim** [reply, marked as answer]
Possible workaround is to not cleanup the emitter right away (delay by max particle life) and set the emit rate to zero

**Comment by kyletcooper** [comment]
Hi @eonarheim thanks for your help, that suggestion seems to have helped the performance issue I was running into!

**Reply by eonarheim** [reply]
Leak fixed! https://github.com/excaliburjs/Excalibur/commit/70d979b27b48fd5f3c8b37a07196275ed74d9c70
